File: src/test/java/com/agiledeveloper/Maps.java
package com.agiledeveloper;

import java.util.*;
import static java.util.stream.Collectors.*;
import static java.util.Comparator.*;

public class Maps {

  public static Map<Integer, List<String>> listByScore(Map<String, Integer> scores) {
    return scores.keySet().stream()
            .collect(groupingBy(scores::get))
            ;
  }

  public static Map<Integer, Set<String>> setByScore(Map<String, Integer> scores) {
      return scores.keySet()
              .stream()
              .collect(groupingBy(scores::get, mapping(String::toUpperCase, toSet())));
  }

  public static Map<Integer, Integer> countScores(Map<String, Integer> scores) {
      return scores.keySet()
              .stream()
              .collect(groupingBy(scores::get, collectingAndThen(counting(), Long::intValue)));
  }

  public static Map<Integer, Set<Integer>> numberOfLetters(Map<String, Integer> scores) {
      return scores.keySet()
              .stream()
              .collect(groupingBy(scores::get, mapping(String::length, toSet())));
  }

  public static Map<Integer, Integer> maxNumberOfLetters(Map<String, Integer> scores) {
      return scores.keySet()
              .stream()
              .collect(groupingBy(
                      scores::get,
                      collectingAndThen(
                              maxBy(comparing(String::length)),
                              name -> name.orElse("").length())))
              ;
  }
}

File: src/test/scala/com/agiledeveloper/Maps.scala
package com.agiledeveloper

object Maps {

  def listByScore(scores: Map[String, Int]): Map[Int, List[String]] =
    scores.keySet
      .groupBy(scores)
      .mapValues(_.toList)

  def setByScore(scores: Map[String, Int]): Map[Int, Set[String]] =
    scores.keySet
      .groupBy(scores)
      .mapValues(
        _.map(_.toUpperCase)
      )

  def countScores(scores: Map[String, Int]): Map[Int, Int] =
    listByScore(scores)
      .mapValues(_.size)

  def numberOfLetters(scores: Map[String, Int]): Map[Int, Set[Int]] =
    listByScore(scores)
      .mapValues(
        _.map(_.size).toSet
      )

  def maxNumberOfLetters(scores: Map[String, Int]): Map[Int, Int] =
    listByScore(scores)
      .mapValues(
        _.foldLeft(0)((top, next) => next.size max top)
      )

}